src/lib/cache_lib/impl.ml

```diff
@@ -127,29 +120,21 @@ module Make (Inputs : Inputs_intf) : Intf.Main.S = struct
       assert_not_consumed t "cannot peek at consumed Cached.t" ;
       value t
 
-    let consume (type a b) (t : (a, b) t) : unit =
-      assert_not_consumed t "cannot consume Cached.t twice" ;
-      mark_consumed t
-
     let transform (type a b) (t : (a, b) t) ~(f : a -> 'c) : ('c, b) t =
-      consume t ;
       attach_finalizer
         (Derivative
            { original= original t
            ; mutant= f (value t)
            ; cache= cache t
-           ; consumed= false
-           ; final_result= final_result t })
+           ; consumed_state= consumed_state t })
 
     let invalidate (type a b) (t : (a, b) t) : a =
-      consume t ;
       mark_failed t ;
       Cache.remove (cache t) (original t) ;
       value t
 
     let free (type a b) (t : (a, b) t) : a =
-      consume t ;
-      mark_applied t ;
+      mark_success t ;
```

Comment:
Can we still call `assert_not_consumed t "cannot consume Cached.t twice" ;` when we try to invalidate cached item that has already been invalidated?